src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs

```diff
@@ -159,6 +188,316 @@ private string GetRootPathWithGuid_NoLock()
             return new MetadataAsSourceFile(fileInfo.TemporaryFilePath, navigateLocation, documentName, documentTooltip);
         }
 
+        private async Task<Document> DecompileSymbolAsync(Document temporaryDocument, OptionSet options, ISymbol symbol, CancellationToken cancellationToken)
+        {
+            // Get the name of the type the symbol is in
+            var containingOrThis = symbol.GetContainingTypeOrThis();
+            var fullName = GetFullReflectionName(containingOrThis);
+
+            var compilation = await temporaryDocument.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            // TODO: retrieve path to actual assembly instead of reference assembly
+            var reference = compilation.GetMetadataReference(symbol.ContainingAssembly);
+            // Load the assembly.
+            var ad = AssemblyDefinition.ReadAssembly(reference.Display, new ReaderParameters() { AssemblyResolver = new RoslynAssemblyResolver(compilation) });
+
+            var decompiler = new CSharpDecompiler(ad.MainModule, GetDecompilerSettings(options));
+            var fullTypeName = new FullTypeName(fullName);
+
+            // Try to decompile; if an exception is thrown the caller will handle it
+            var text = decompiler.DecompileTypeAsString(fullTypeName);
+            return temporaryDocument.WithText(SourceText.From(text));
+        }
+
+        private static DecompilerSettings GetDecompilerSettings(OptionSet options)
+        {
+            var decompilerSettings = new DecompilerSettings();
+            var formattingOptions = decompilerSettings.CSharpFormattingOptions;
+
+            // Indentation
+            formattingOptions.IndentNamespaceBody = true;
```

Comment:
consider using an ObjectInitializer for this.

Reply:
➡️ API does not provide a constructor for this type

Reply:
?   DecompilerSettings is instantiated above...

Reply:
i'm not getting why you can't do:

```var settings = new DecompilerSettings {
  CSharpFormattingOptions = { 
      IndentNamespaceBody = true ...
  }
};
```

Reply:
💭 Technically allowed, but I find the benefits questionable. I don't find it better or worse necessarily, but it's less work to not change it so that become my preference.

Reply:
SGTM